packages/validator/src/services/doppelgangerService.ts

```diff
@@ -0,0 +1,232 @@
+          indicesToCheck.push(index);
+          indicesToCheckMap.set(index, pubkeyHex);
+        }
+      }
+    }
+
+    if (indicesToCheck.length === 0) {
+      return;
+    }
+
+    // in the current epoch also request for liveness check for past epoch in case a validator index was live
+    // in the remaining 25% of the last slot of the previous epoch
+    const [previous, current] = await Promise.all([
+      this.getLiveness(indicesToCheck, currentEpoch - 1),
+      this.getLiveness(indicesToCheck, currentEpoch),
+    ]);
+
+    this.detectDoppelganger(currentEpoch, previous, current, indicesToCheckMap);
+  };
+
+  private async getLiveness(indicesToCheck: ValidatorIndex[], epoch: Epoch): Promise<LivenessResponseData[]> {
+    if (epoch < 0) {
+      return [];
+    }
+
+    try {
+      const {data} = await this.api.validator.getLiveness(indicesToCheck, epoch);
+      return data;
+    } catch (e) {
+      this.logger.error(`Error getting liveness data for epoch ${epoch}`, {}, e as Error);
+      return [];
+    }
+  }
+
+  private detectDoppelganger(
+    currentEpoch: Epoch,
+    previousEpochLiveness: LivenessResponseData[],
+    currentEpochLiveness: LivenessResponseData[],
+    indicesToCheckMap: Map<ValidatorIndex, PubkeyHex>
+  ): void {
+    const previousEpoch = currentEpoch - 1;
+    const violators: ValidatorIndex[] = [];
+
+    // Perform a loop through the current and previous epoch responses and detect any violators.
+    //
+    // A following loop will update the states of each validator, depending on whether or not
+    // any violators were detected here.
+
+    for (const responses of [previousEpochLiveness, currentEpochLiveness]) {
+      for (const response of responses) {
+        if (!response.isLive) {
+          continue;
+        }
+
+        const state = this.doppelgangerStateByPubkey.get(indicesToCheckMap.get(response.index) ?? "");
+        if (!state) {
+          this.logger.error(`Inconsistent livenessResponseData unknown index ${response.index}`);
+          continue;
+        }
+
+        if (state.nextEpochToCheck <= response.epoch) {
+          // Doppleganger detected
+          violators.push(response.index);
+        }
+      }
+    }
+
+    if (violators.length > 0) {
+      // If a single doppelganger is detected, enable doppelganger checks on all validators forever
+      for (const state of this.doppelgangerStateByPubkey.values()) {
+        state.remainingEpochs = Infinity;
+      }
+
+      this.logger.error(
+        "Doppelganger(s) detected\n" +
+          "A doppelganger occurs when two different validator clients run the same public key. \
+          This validator client detected another instance of a local validator on the network \
+          and is shutting down to prevent potential slashable offences. Ensure that you are not \
+          running a duplicate or overlapping validator client",
+        violators
+      );
+
+      // TODO: Request process to shutdown
```

Comment:
what's the plan for this TODO?

Reply:
Will implement in a latter PR.

Current design prevents any validator from signing slash-able messages forever. Killing the process is an important signal to node operators but not critical for base functionality.